From d174a16ea0564b5206e3675ad10affa1f225211a Mon Sep 17 00:00:00 2001
From: Liu Linhong <linhong.liu@intel.com>
Date: Wed, 11 Mar 2015 17:54:54 +0800
Subject: [PATCH] Add otg descriptor to usb composite dev desc

In USB B-OTG compliance test, host side need read otg descriptor
from DUT's composite dev descriptor to enumerate DUT.

The compliance will get otg descriptor twice. This is the second
time.

Change-Id: Idb10fb0ccf8698e1637489334700dbcc28c2231b
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-9620
Signed-off-by: Liu Linhong <linhong.liu@intel.com>
---
 drivers/usb/gadget/android.c   | 3 +++
 drivers/usb/gadget/composite.c | 6 ++++++
 include/linux/usb/composite.h  | 3 +++
 3 files changed, 12 insertions(+)

diff --git a/drivers/usb/gadget/android.c b/drivers/usb/gadget/android.c
index fc090fcfcb43..a3a4d7c330b3 100644
--- a/drivers/usb/gadget/android.c
+++ b/drivers/usb/gadget/android.c
@@ -1550,6 +1550,9 @@ static int android_bind(struct usb_composite_dev *cdev)
 	device_desc.iSerialNumber = id;
 	dev->reset_string_id = id;
 
+	if (gadget_is_otg(gadget))
+		cdev->otg_desc = &otg_descriptor;
+
 	usb_gadget_set_selfpowered(gadget);
 	dev->cdev = cdev;
 
diff --git a/drivers/usb/gadget/composite.c b/drivers/usb/gadget/composite.c
index 8c9bbc855d09..d2a7e99b299a 100644
--- a/drivers/usb/gadget/composite.c
+++ b/drivers/usb/gadget/composite.c
@@ -1310,6 +1310,12 @@ composite_setup(struct usb_gadget *gadget, const struct usb_ctrlrequest *ctrl)
 				value = min(w_length, (u16) value);
 			}
 			break;
+		case USB_DT_OTG:
+			if (cdev->otg_desc) {
+				memcpy(req->buf, cdev->otg_desc, w_length);
+				value = w_length;
+			}
+			break;
 		}
 		break;
 
diff --git a/include/linux/usb/composite.h b/include/linux/usb/composite.h
index a24ac8dadaa1..5d2a6f234979 100644
--- a/include/linux/usb/composite.h
+++ b/include/linux/usb/composite.h
@@ -394,6 +394,9 @@ struct usb_composite_dev {
 
 	/* protects deactivations and delayed_status counts*/
 	spinlock_t			lock;
+
+	/* OTG support */
+	struct usb_otg_descriptor	*otg_desc;
 };
 
 extern int usb_string_id(struct usb_composite_dev *c);
-- 
2.1.0

